File: python/pig_latin.py
def aux(text):
    vowels = 'aeiou'

    # Rule 1
    if text[0] in vowels:
        return text
    if text[:2] in ('xr', 'yt'):
        return text

    # Rule 3
    i = text.find('qu')
    if i != -1:
        return text[i+2:] + text[:i+2]

    # Rule 2, 4
    i = 0
    while text[i] not in vowels:
        if i > 0 and text[i] == 'y':
            break
        i += 1
        
    return text[i:] + text[:i]


def translate(text):
    arr = text.split(' ')
    return ' '.join([aux(s) + 'ay' for s in arr])

File: dart/pig_latin.dart
bool isVowel(String s) {
  return RegExp(r'[aeiou]').hasMatch(s);
}

String aux(String s) {
  if (isVowel(s[0])) {
    return s;
  }

  if (s.substring(0, 2) == 'xr' || s.substring(0, 2) == 'yt') {
    return s;
  }

  int i = s.indexOf('qu');
  if (i != -1) {
    return s.substring(i + 2) + s.substring(0, i + 2);
  }

  i = 0;
  while (!isVowel(s[i])) {
    if (i > 0 && s[i] == 'y') {
      break;
    }
    i++;
  }
  return s.substring(i) + s.substring(0, i);
}

String translate(String s) {
  return s.split(' ').map((e) => aux(e) + 'ay').join(' ');
}
